crates/valence/src/equipment.rs

```diff
@@ -0,0 +1,240 @@
+            } else {
+                None
+            }
+        })
+    }
+
+    fn set_modified_slot(&mut self, slot: EquipmentSlot) {
+        let shifts: i8 = slot.into();
+        self.modified_slots |= 1 << (shifts as u8);
+    }
+
+    fn clear_modified_slot(&mut self) {
+        self.modified_slots = 0;
+    }
+}
+
+impl TryFrom<i8> for EquipmentSlot {
+    type Error = &'static str;
+
+    /// Convert from `id` according to https://wiki.vg/Protocol#Set_Equipment
+    fn try_from(id: i8) -> Result<Self, Self::Error> {
+        let slot = match id {
+            0 => EquipmentSlot::MainHand,
+            1 => EquipmentSlot::OffHand,
+            2 => EquipmentSlot::Boots,
+            3 => EquipmentSlot::Leggings,
+            4 => EquipmentSlot::Chestplate,
+            5 => EquipmentSlot::Helmet,
+            _ => return Err("Invalid value"),
+        };
+
+        Ok(slot)
+    }
+}
+
+impl From<EquipmentSlot> for i8 {
+    /// Convert to `id` according to https://wiki.vg/Protocol#Set_Equipment
+    fn from(slot: EquipmentSlot) -> Self {
+        match slot {
+            EquipmentSlot::MainHand => 0,
+            EquipmentSlot::OffHand => 1,
+            EquipmentSlot::Boots => 2,
+            EquipmentSlot::Leggings => 3,
+            EquipmentSlot::Chestplate => 4,
+            EquipmentSlot::Helmet => 5,
+        }
+    }
+}
+
+/// When a [Equipments] component is changed, send [SetEquipment] packet to all clients
+/// that have the updated entity in their view distance.
+///
+/// NOTE: [SetEquipment] packet only have cosmetic effect, which means it does not affect armor resistance or damage.
+pub fn update_equipment(
+    mut equiped_entities: Query<(Entity, &McEntity, &mut Equipments), Changed<Equipments>>,
+    mut clients: Query<(Entity, &mut Client)>,
+) {
+    for (equiped_entity, equiped_mc_entity, mut equips) in &mut equiped_entities {
+        if !equips.has_modified_slots() {
+            continue;
+        }
+
+        let equiped_instance = equiped_mc_entity.instance();
+        let equiped_chunk_pos = ChunkPos::from_dvec3(equiped_mc_entity.position());
+
+        // Send packets to all clients which have the equiped entity in view
+        for (client_entity, mut client) in &mut clients {
+            if client.instance() != equiped_instance {
+                continue;
+            }
+
+            // It is not necessary to `SetEquipment` packets for the associate client's player entity,
+            // because its equipments are already tracked on the client side.
+            if client_entity != equiped_entity && client.view().contains(equiped_chunk_pos) {
+                client.write_packet(&SetEquipment {
+                    entity_id: VarInt(equiped_mc_entity.protocol_id()),
+                    equipment: equips.iter_modified_equipments().collect(),
+                });
+            }
+        }
+
+        equips.clear_modified_slot();
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
This doesn't have any tests that assert for sent packets. See the examples here for how to set this up: https://github.com/valence-rs/valence/blob/main/crates/valence/src/unit_test/example.rs